Add amd64_llvm platform, using the LLVM assembler

mokomull opened the pull request:
This builds against LLVM 3.8.  I also have commits from earlier attempts at this for LLVM 3.4 and 3.6 if you would prefer one of those — I picked 3.8 since it's available on both Ubuntu 14.04 and 16.04.

This builds with `make LLVM_CONFIG=llvm-config-3.8 ARCH=amd64_llvm`.  After doing that, `bin/rappel` doesn't fork nasm, but instead uses the LLVM infrastructure in-process.  Note: LLVM uses AT&T syntax, unlike nasm.

This probably isn't finalized; as I typed out the above, I will probably add that to the README and refresh this PR later.

I am licensing my contributions under the 2-clause BSD to give you the opportunity to drop the advertising clause at a later date.

mokomull committed:
Add amd64_llvm platform, using the LLVM assembler

This builds against LLVM 3.8.

As far as I can tell, memory allocations are freed -- LLVM has really weird
patterns for memory use.  This took some time to track down exactly which
pointers I'm expected to delete and which ones I'm not.

  - Most of MCContext's get___ methods use a custom allocator, so they do not
    need to be deleted.
  - The MCObjectStreamer creates an internal MCAssembler with some of the
    constructor arguments I give it.  It reaches into that at ~MCObjectStreamer
    time and deletes them out from under me.

mokomull commented:
Maybe this code would've been cleaner if I'd seen #6 first.  I'll have to check that out.

yrp604 commented:
Interesting, thank you very much for the PR.

Aside from avoiding the overhead of a fork()/wait() for every assembly, what are the main benefits LLVM provides? AT&T syntax?

This, or something similar, was my original plan for rappel and would remove some cruft (nasm doesn't play well with seekable input resulting in the creation of tmp files), but I'm kinda hesitant to exchange nasm (which is pretty lightweight and self contained) with LLVM which is significantly larger and has a higher ongoing maintenance cost (especially since my C++ is at best, quite poor.)

mokomull commented:
All of the above are reasons I added it as a separate `arch/` subdirectory — the syntax change alone is a large enough difference in behavior that I didn't feel comfortable completely replacing `arch/amd64`.

I think using in-process avoids writing-out the user input to a file as well, but at the cost of having added an ELF parser — all I have found so far in LLVM creates ELF structures in memory (for a *nix target, at least).

Maintenance-wise, the conversion from 3.4→3.6→3.8 have been fairly straightforward.  I intend to keep maintaining this for the foreseeable future, since I've found the tool incredibly useful, especially while trying to make sense out of the Intel documentation :)

yrp604 commented:
Glad you've found it useful! If you have any other feature requests, I'm all ears.

My thinking right now is if I merge it to have it #ifdef'd off and enabled via a Makefile target, then allowed it to be selected at runtime. Although, as you point out, I think keystone might provide a more consistent api for various architectures while using the same backend? I've only tried the python backend for keystone, and it was pretty straightforward (although at the expense of another dep).

One of the goals I had in mind was to keep rappel relatively light dependency wise, so adding LLVM (or LLVM/keystone) to the default install is a bit less appealing. Also, the kinda-hacky way I define functions to their architecture equiv makes selecting it at runtime kinda difficult... If I did that my guess is I would throw it on another branch, but to be honest I don't think enough people use the project to get meaningful feedback on nasm/as vs LLVM/keystone (let alone realize the other branch exists).

mokomull commented:
The `Makefile` changes I included require you to set `ARCH=amd64_llvm` manually — the auto-detection via `uname` would still build the existing `arch/amd64` (which is unmodified in this PR).

I'll definitely update the readme to make this more clear and update my branch.

mokomull committed:
llvm: use llvm-config-3.8 if unspecified

This only affects builds with `ARCH=amd64_llvm`, and the user is still able to
specify a path to a particular `llvm-config`.  Adding this default reduces the
burden of determining which LLVM API version is currently used.

mokomull committed:
README: add documentation for using LLVM support

mokomull commented:
I updated the documentation to clarify that this is a compile-time choice.  If you'd prefer the ability to compile both backends and select at runtime, I've got an idea there — likely, add a function pointer to the global options struct.

yrp604 commented on arch/amd64_llvm/assemble_intel.cpp:
```diff
@@ -0,0 +1,216 @@
+		llvm::TargetRegistry::lookupTarget(LLVM_HOST_TRIPLE, error);
+	if (!target) {
+		std::cerr << "Could not find target: " << error << std::endl;
+		abort();
```
EXIT_FAILURE?

mokomull replied:
Normally I prefer `abort()`, but you're right—that doesn't seem to help any here.  Will update to use `exit(EXIT_FAILURE)` like the rest of the code.

yrp604 commented:
Cool, this looks good. Can you just add the relevant files/packages that this requires to the readme? I assume they're in some llvm dev deb/rpms.

mokomull commented:
Updated the commits to incorporate review feedback and add documentation.

Unfortunately, I found that Debian does not have LLVM 3.8 available at all, even in sid.  The packages I referenced are available in Ubuntu 14.04 and 16.04.

yrp604 commented:
Awesome, thanks. I'll do a final pass this weekend and should have it merged soon. Thanks, looking forward to playing with this.

yrp604 commented:
So, first apologies for the delay in finishing this.

After some consideration, as well as building and poking at it, I'm less inclined to merge this. The code looks great, and works exactly as advertised, I'm just back at the maintenance issue. As evidenced by the delay in replying to this issue, I don't really think, work on, or modify rappel very frequently. Basically when I personally need a new feature I'll gin it up. The unfortunate side effect of this is when I don't need new features, I'm not very inclined to pay attention to this project.

I more or less see two solutions to this:
1) I add people with commit access to the project allowing them to maintain the code over time. I dislike this, because as I mentioned above this project is basically my debugger swiss army knife in many respects, and the ability to quickly modify it is important to me (without reading other people's changes).
2) I link to and encourage forks of the project.

So, with that somewhat long winded explanation, would you like a link from rappels README to your changes? I apologies for wasting your time in previous comments. I really appreciate the time and changes you made, and hope you understand my reluctance to merge them.

mokomull commented:
Somehow, I completely missed the previous update, but I do agree with your assessment.  I'll continue to maintain this on the side.  Thanks for your guidance!
